File: app/src/main/java/com/example/stas/oxforddictionary/presentation/presenter/entry/EntryPresenter.java
package com.example.stas.oxforddictionary.presentation.presenter.entry;

import com.example.stas.oxforddictionary.domain.model.definition.DefinitionResult;
import com.example.stas.oxforddictionary.presentation.mapper.definition.DefinitionModelMapperKt;
import com.example.stas.oxforddictionary.presentation.view.entry.adapter.Item;
import com.example.stas.oxforddictionary.domain.interactor.DefinitonInteractor;
import com.example.stas.oxforddictionary.presentation.mapper.definition.DefinitionModelDataMapper;
import com.example.stas.oxforddictionary.presentation.view.base.ErrorHandler;
import com.example.stas.oxforddictionary.presentation.view.entry.EntryContract;
import com.example.stas.oxforddictionary.presentation.viewmodel.definition.LexicalEntryModel;
import com.example.stas.oxforddictionary.presentation.viewmodel.definition.PronunciationModel;
import com.example.stas.oxforddictionary.presentation.viewmodel.definition.ResultModel;
import com.example.stas.oxforddictionary.presentation.viewmodel.definition.SenseModel;
import com.example.stas.oxforddictionary.presentation.viewmodel.definition.SubsenseModel;

import java.util.ArrayList;
import java.util.List;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;

public class EntryPresenter implements EntryContract.Presenter {
    private EntryContract.View view;
    private final DefinitonInteractor interactor;
    private final CompositeDisposable compositeDisposable;
    private final ErrorHandler errorHandler;
    private final DefinitionModelDataMapper definitionModelDataMapper;

    public EntryPresenter(DefinitonInteractor interactor, CompositeDisposable compositeDisposable,
                          ErrorHandler errorHandler, DefinitionModelDataMapper definitionModelDataMapper) {
        this.interactor = interactor;
        this.compositeDisposable = compositeDisposable;
        this.errorHandler = errorHandler;
        this.definitionModelDataMapper = definitionModelDataMapper;
    }

    @Override
    public void attachView(EntryContract.View view) {
        this.view = view;
        errorHandler.attachView(this.view);
    }

    @Override
    public void detachView(){
        errorHandler.detachView();
        view = null;
        compositeDisposable.dispose();
    }

    @Override
    public void getDefinition(String word) {
        Disposable definitionDisp = interactor.loadDefinition(word)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Consumer<DefinitionResult>() {
                    @Override
                    public void accept(DefinitionResult result){
                        ResultModel resultModel = DefinitionModelMapperKt.toViewModel(result);
                        view.showDefinition(extractDefinitions(resultModel), extractTitle(resultModel));
                    }
                }, new Consumer<Throwable>() {
                    @Override
                    public void accept(Throwable throwable) {
                        view.hideProgressBar();
                        errorHandler.proceed(throwable);
                    }
                });
        compositeDisposable.add(definitionDisp);
    }

    @Override
    public void getSound(String word) {
        Disposable soundDisp = interactor.loadDefinition(word)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Consumer<DefinitionResult>() {
                    @Override
                    public void accept(DefinitionResult result){
                        view.playSound(result.getLexicalEntries().get(0).getPronunciations().get(0).getAudioFile());
                    }
                }, new Consumer<Throwable>() {
                    @Override
                    public void accept(Throwable throwable) {
                        errorHandler.proceed(throwable);
                    }
                });
        compositeDisposable.add(soundDisp);
    }

    private List<Item> extractDefinitions(ResultModel result){
        List<Item> definitions = new ArrayList<>();
        for (LexicalEntryModel lexicalEntry : result.getLexicalEntries()) {
            definitions.add(lexicalEntry);
            for(SenseModel sense : lexicalEntry.getEntries().get(0).getSenses()){
                if(!sense.getDefinitions().isEmpty()){
                    definitions.add(sense);
                    for (SubsenseModel subsense :sense.getSubsenses()) {
                        if(!subsense.getDefinitions().isEmpty()){
                            definitions.add(subsense);
                        }
                    }
                }
            }
        }
        return definitions;
    }
    private List<String> extractTitle(ResultModel result){
        LexicalEntryModel lexicalEntry = result.getLexicalEntries().get(0);
        List<String> titleSet = new ArrayList<>();
        titleSet.add(lexicalEntry.getText());
        for (PronunciationModel pronunciation: lexicalEntry.getPronunciations()){
            if (pronunciation.getAudioFile() != null){
                titleSet.add("[" + pronunciation.getPhoneticSpelling()+ "]");
            }
        }
        return titleSet;
    }
}

File: app/src/main/java/com/example/stas/oxforddictionary/presentation/presenter/entry/EntryPresenter.kt
package com.example.stas.oxforddictionary.presentation.presenter.entry

import com.example.stas.oxforddictionary.data.database.model.SavedWordModel
import com.example.stas.oxforddictionary.data.database.model.ViewedWordModel
import com.example.stas.oxforddictionary.domain.usecase.DefinitonInteractor
import com.example.stas.oxforddictionary.domain.usecase.definition.SaveToViewedUseCase
import com.example.stas.oxforddictionary.domain.usecase.definition.SaveWordUseCase
import com.example.stas.oxforddictionary.domain.usecase.definition.SearchViewedWordsUseCase
import com.example.stas.oxforddictionary.presentation.mapper.definition.toViewModel
import com.example.stas.oxforddictionary.presentation.view.base.ErrorHandler
import com.example.stas.oxforddictionary.presentation.view.entry.EntryContract
import com.example.stas.oxforddictionary.presentation.view.entry.adapter.Item
import com.example.stas.oxforddictionary.presentation.viewmodel.definition.ResultModel
import io.reactivex.Observable
import io.reactivex.ObservableSource
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import io.reactivex.subjects.PublishSubject
import java.util.*
import java.util.concurrent.TimeUnit


class EntryPresenter(private val interactor: DefinitonInteractor, private val compositeDisposable: CompositeDisposable,
                     private val errorHandler: ErrorHandler, private val saveWordUseCase: SaveWordUseCase,
                     private val saveToViewedUseCase: SaveToViewedUseCase,
                     private val searchViewedWordsUseCase: SearchViewedWordsUseCase) : EntryContract.Presenter {
    private var view: EntryContract.View? = null
    private var subject: PublishSubject<String>? = null

    override fun attachView(view: EntryContract.View) {
        this.view = view
        errorHandler.attachView(view)
        subject = PublishSubject.create<String>()
    }

    override fun detachView() {
        errorHandler.detachView()
        view = null
        compositeDisposable.dispose()
    }

    override fun getDefinition(word: String) {
        val definitionDisp = interactor.loadDefinition(word)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ result ->
                    val resultModel = result.toViewModel()
                    view!!.showDefinition(extractDefinitions(resultModel), extractTitle(resultModel))
                    saveToViewed(word)
                }, { throwable ->
                    view!!.hideProgressBar()
                    errorHandler.proceed(throwable)
                })
        compositeDisposable.add(definitionDisp)
    }

    private fun saveToViewed(word: String) {
        val saveViewedDisp = saveToViewedUseCase.saveWord(ViewedWordModel(word = word))
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ view?.showError("Saved") },
                        { throwable -> errorHandler.proceed(throwable) })
        compositeDisposable.addAll(saveViewedDisp)
    }

    override fun getSound(word: String) {
        val soundDisp = interactor.loadDefinition(word)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ (_, _, lexicalEntries) ->
                    view!!.playSound(lexicalEntries[0].pronunciations[0].audioFile)
                }, { throwable -> errorHandler.proceed(throwable) })
        compositeDisposable.add(soundDisp)
    }

    override fun saveDefinition(word: String, definition: String) {
        val saveWordDisp = saveWordUseCase.saveWord(SavedWordModel(value = word, definition = definition, guess = 0))
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({
                    view?.showError("DONE!")
                },
                        { throwable ->
                            errorHandler.proceed(throwable)
                        })
        compositeDisposable.add(saveWordDisp)
    }

    override fun onEntryTextChanged(text: String) = searchViewedWordsUseCase.searchViewedWords("$text%")

    private fun extractDefinitions(result: ResultModel): List<Item> {
        val definitions = ArrayList<Item>()
        for (lexicalEntry in result.lexicalEntries) {
            definitions.add(lexicalEntry)
            for (sense in lexicalEntry.entries[0].senses) {
                if (!sense.definitions!!.isEmpty()) {
                    definitions.add(sense)
                    for (subsense in sense.subsenses) {
                        if (!subsense.definitions!!.isEmpty()) {
                            definitions.add(subsense)
                        }
                    }
                }
            }
        }
        return definitions
    }

    private fun extractTitle(result: ResultModel): List<String?> {
        val (_, _, _, pronunciations, text) = result.lexicalEntries[0]
        val titleSet = ArrayList<String?>()
        titleSet.add(text)
        for ((audioFile, _, _, phoneticSpelling) in pronunciations) {
            if (audioFile != null) {
                titleSet.add("[$phoneticSpelling]")
            }
        }
        return titleSet
    }
}

